### ISSUE TITLE
Move non-essential java.security code out of core Scala Native

### ISSUE THREAD

**Issue by ekrich** [scala-native/scala-native#2601]
Edit: There is also one `javax.security` class that needs to be moved out too.

Refer to https://github.com/scala-native/scala-native/issues/2600

And the current potential target. https://github.com/lolgab/scala-native-crypto

**Comment by LeeTibbert** [comment]
Recent discussions about TLS brought me here.  I think we all concur that since
this involves removing a big chunk, it should be done in a 0.x.0 version change.

I think this Issue is not a Blocker, but only a few levels below that.   At least
as far as having a plan to do it in 0.6.0, and not 0.5.0.

IIRC, the situation with third-party library providing `java.security` definitions
comes up when it tries to provide definitions which are already in SN `java.security`,
especially when the existing definition does not work in the 3rd party library.

I believe that a hypothetical 3rd party `javax.net.ssl` requires at least 4 classes
in `java.security`.  It remains to be shown if they can be implemented as a 3rd
party overlay to the current scheme.

Now, I would not invest resources in moving all of SN `java.security` to a 
(theoretical) 3rd party because of TLS support which may or may happen.
I do think that there should be a short, explicit plan about "What to do about Bob?".
If only the plan is "Do not add anything additional to SN `java.security` and wait for
0.6.0.  In the "Do not add" case, it would be nice to put a README statemment of
intent in that directory so that both SN devos and application devos know not to expect
 anything more there.  (I believe that Java 17 `java.security` still exists and is supported).

**Comment by ekrich** [comment]
Scala.js has `Guard`, `Permission`, and `Throwables` which has `AccessControlException`. Much of the removing code has to do with lessons learned by Scala.js. We can find out the rationale.

**Comment by armanbilge** [comment]
Indeed, many of the Scala Native implementations of `java.security` appear to be incomplete. For example:
- https://github.com/scala-native/scala-native/blob/672b4a9932443df1a351decd05ea6bdf0357d1f1/javalib/src/main/scala/java/security/MessageDigest.scala
- https://github.com/scala-native/scala-native/tree/main/javalib/src/main/scala/java/security/cert

A good rule of thumb is, if you don't want to implement it in its entirety, then don't include it in core, because it will prevent someone else from doing so.

**Comment by ekrich** [comment]
We have the opportunity to remove this in the 0.5.0 release so maybe we need to at least bite that off now since everyone is going to rebuild for 0.5.0. Where would it land, in the Scala Native org?

**Comment by LeeTibbert** [comment]
Yes, asking  asking  "What would SJS do (a second time)?" is worthwhile. 

 As mentioned above, I think the over-arching issue is getting something (acceptable, we are 
going to have to live with this a __long__ time) moving soon so that it
catches the 0.5.0 train.     To my thinking, other considerations are at a lessor level of urgency (as 
long as you pick my opinions ;-))

To be fair, I should state that my current opinion is that `java.security` should be
removed `root & branch`. 

We all probably concur that decisions that endure and are only reversed with difficulty should be made on
 the basis of actual information.  

1. At the next level down, I see the need for a focused "design study" to see if all of
`java.security` can be removed or if some needs to remain for the use of SN core.
This needs to be considered in the light of Java 8, how it has evolved to Java 17,
and Java's announced plans for the future

2. The question of "Where will it live" is the next level of urgency.  It would be nice if it could live
somewhere under "scala native" or "scala" itself.   No one is going to want to use "mom&pop", "trustme",
or "iAmNotHiddenStateActor" security modules.

3. There is a good reason to believe that a reasonable
    course of action might be "Announce in 0.5.0 that,
    absent severe errors, no more changes will be made
    in `java.security`. No new classes.  The plans is
    to remove these classes and recommend 3rd party
    implementations in 0.6.0". A similar "No intent to implement
    `javax.net` or `javax.net.ssl`, recommed 3rd party implementations"

    I think that preserves the possibility of  "3rd party"
    `java.security` and, especially, `javax.net.ssl` implementions.
    The former could not have classes which currently exist
    in `java.security` until those are removed (i.e. 0.6.0).
    That would give time to provide an alternate home.

I believe that this issue has strong merit on its own, independent
of any `javax.net.ssl` discussion going on elsewhere.

**Comment by armanbilge** [comment]
My opinion is that every class in `java.security` should either be:
1. Removed in 0.5.0; or
2. Implemented fully and correctly in 0.5.0, or at least explicitly promised that it will be in some 0.5.x release.

The reason is simple: if neither of those are true, then it means it will be impossible to have a working version of those classes until 0.6.0. I think that's an unacceptable blocker.

**Comment by LeeTibbert** [comment]
> Removed in 0.5.0

I concur with the idea/drive/intent  but the implementation side of me would substitute 'timely & urgent'. 

Beyond arriving at design consensus, the major time consuming items are 
1)  finding a timely solution to SN internal use
2) finding someone to bell the cat.

As a "(very) nice to have":
It would be good if there were at least one 3rd party replacement implementation.
If there _are_ existing users, having such would reduce the grumbles.
(or at least documenting how to build one's own jar for these classes).
A full 3rd party implementation  is a time consuming task.  
I sincerely hope that 0.5.0 is out before then.  The "documenting
with caveats minimal roll-your-own-of-evicted-methods) solution is not as nice
could possibly make 0.5.0). Notice that I am not listing this as a blocker to a
0.5.0 change.

> Implemented fully and correctly in 0.5.0,  or at least explicitly promised that it will be in some 0.5.x release.

Doing a quick find & grep of files not in the `java.security` package importing `java.security`
I found three uses. I think that lolgab might be using MessageDigest for his privately
published implementation. If so, that would be at least one user 'in the wild'.

As a thought experiment in possibilities, files other than those use in SN itself could be removed
immediately in the 0.5.0 stream to reduce the "exposure surface" and allow time to study
how best to handle the files in active use. 

If a given `java.security` class is used in only one package, the former could be made private to the latter; 
sealed with the Curse of the Mummy". Some of the `java.time` stuff may have used a `javalib-ext` mechanism;
I was offline when that happened and have not caught up.  Same issues, though, if we are going to 
live with it, how do we know it is good(secure?).

Do we have any evidence that those classes have defects?  As a counter, do we
have any evidence, beyond passing, probably minimal, unit-tests of unknown quality 
that they are fit for purpose?    I have no evidence either way, beyond that the
MessageDigest declaration (I think the class is "dummy" has been used without report).

Of course, any reports of defects should be addressed immediately.

`UUID.scala`, of recent discussion in another thread, is in `java.util`.  We concur
on the urgency of a fix there.

> Implemented fully and correctly in 0.5.0, or at least explicitly promised that it will be in some 0.5.x release.
Is there anybody who could make such a promise with the likelihood that it will be fulfilled?  

As to the `Implemented fully` part. I would have to check the Java Doc to see the full extent of what is in
and under `java.security`.  I suspect it is huge and that each of the classes could take infinite resources
to implement & maintain.  Anyone having the knowledge & skill probably has a job.

I personally find the 'spin it off to a 3rd party' approach distasteful but bow to its exigency.

**Comment by ekrich** [comment]
I think it would be useful to publish what we have now as a junk `java.security` package or wrap the code into an existing project so that if people are depending on the classes they have a place to go to not break their app or lib when `0.5.0` rolls out.

I know from personal experience what it is like to get the `java.time` stuff ripped out right before the `0.4.0` release and have no alternative which also kept everything downstream from my lib from updating to the new version of Scala Native. I had to spend a lot of time porting a library. Now there is no interest in that lib because of a better alternative so it was mostly wasted time.

**Comment by LeeTibbert** [comment]
I mostly concur.
>  `junk java.security`  Two words which, side by each, should cause angina & terror.

"Internal usage", "What to do about 0.4.x", and "Finding a home" are Big Stones.

I think the "Internal usage" can be figured out if dedicated time (a week or two) is spent
on it.  

0.4.x deserves careful consideration.  Depends on if SN has any identified users of that
code. Looping the one known user in early would be a good idea.  If there are 0 know (until
we break it) uses, it makes things simpler.  If 0.4.x methods are removed, we would be
able to discuss security concerns directly with identified, affected users.  Breaking
things on people is a __last_extreme_.   

With other code, I would suggesting moving it to `java.ugly.useAtYourOwnRisk.foo`. 
Leaving known, or suspected, bad (as opposed to incomplete) security code around
 

Do we have any hard evidence on how bad, beyond incomplete, existing 0.4.x `java.security` code is?

> so that if people are depending on the classes they have a place to go to not break their app or lib when 0.5.0 rolls out
Yes, I will sing in that choir.

**Comment by LeeTibbert** [comment]
Fun keeps coming: 

I just discovered `/javax/security/auth/x500/X500Principal.scala`  so
`javax/security` should be part of the discussion.

It looks like that file has no current internal users. Sick to say, 
but this is one of the times where having no unit-test for a feature
pays off.  (In case anyone misses the irony, I do not mean a word
of that.)

**Comment by LeeTibbert** [comment]
To size/scope the `java.security effort`:

Everything needs to be looked at by (two or more) sets
of eyes before signing off on for 0.4.x (or 0.5.x)

These look hard to get wrong.

```
AccessControlException.scala
GeneralSecurityException.scala
NoSuchAlgorithmException.scala

Principal.scala // used in a number of places. It is pretty minimal.
```

Need code review & explicit exoneration.
```
/cert/*
CodeSigner.scala

MessageDigest.scala  // What does lolgab use?  Who in git blame?
MessageDigestSpi.scala // What does lolgab use? Who in git blame?

Timestamp.scala
```

**Comment by LeeTibbert** [comment]
Title should probably be changed to include `javax.security`.

**Comment by ekrich** [comment]
I already edited in the first comment.

**Comment by armanbilge** [comment]
@LeeTibbert regarding this:
```
MessageDigest.scala  // What does lolgab use?  Who in git blame?
MessageDigestSpi.scala // What does lolgab use? Who in git blame?
```

I can't speak for other uses, but Lorenzo has actually _reimplemented_ those classes:
https://github.com/lolgab/scala-native-crypto/blob/c45268b5787030a4624ae1191e4bc2e4438eba37/scala-native-crypto/src/java/security/MessageDigest.scala

So I wouldn't say "used" ... I would say replaced. Because the existing implementation in Scala Native is not complete enough to be useful.

**Comment by LeeTibbert** [comment]
@armanbilge  re: lolgab   Understood that he re-implemented.  I have see the file names but did not
have time to check the contents to see if he may used something in the SN `java.util`.  My comment
was more of a placeholder for the need for further checking, to avoid hosing someone.

**Comment by ekrich** [comment]
Note: some `java.security` classes are used in Java jar/zip code.

**Comment by LeeTibbert** [comment]
Eric.  Much as I like the conceptual simplicity of "move all", I suggest "assess" or "triage"
is a better description/goal.

During the course of other work, I have been flitting into various files in `java.security`.
Your expressed wish to expunge them was present in my mind.  As I looked at them
piecemeal, I saw nothing that was a security thread.

Now that is a ragged impression.  If I were writing the story, I would say
the first pass would be to develop a list of modules, definitions, etc currently
in `java.security`.  The write out two checklists based on those names.
Ask two people to accept a checklist and, independently, review the
modules and mark them as harmless or possible threat or security issue.
Then compare the two lists.  

I for one would not mark an abstract class for eviction, unless
it was the last file in the package.

The background for all of this is that to the best of my knowledge
there is no third party replacement at present. I think there
is not a prayer of a third party replacement in the next eon or two.
The argument for eliminating 'innocent' modules would be stronger
if there were a replacement. 

Well, there is always the replacement of "cut & paste the old published code".
That works for 'innocent' but not potentially harmful modules.

Yes, every module in SN must earn its keep, best by having an
identified current user or champion.   Modules identified as
security (CERT) risks or a danger to execution stability need
to be fixed or extirpated.    

Assuming that everything in `java.security` is in the latter
class is a pretty large leap.

**Comment by LeeTibbert** [comment]
I am not the driver of this bus and do not want to introduce dis-harmony or dissension.
There is too much to be done and too few of us to do it to get bogged down in 
Jonathan Swift issues.

If I were in a meeting asking how available time might be mapped to requests, I would
say that a review of `java.security` was not the Highest priority but the priority below.
If that review reported no critical issue, I would rank removing it as the lowest priority.
That is, if someone were motivated to do the removal and, especially, document
both the removal and workarounds, then this is a contributory community, let them
have at it, but do not ask people to do it.

**Comment by ekrich** [comment]
We do have this repo for putting the old code to help with `0.5.0` migration: https://github.com/scala-native/scala-native-java-stubs

Unfortunately, as @LeeTibbert has done some of the homework, it will take some work to figure out everything here. I think the goal should be to remove all the code we can that we don't want to maintain and match Scala.js as much as possible. I changed the title to match this intent.

**Comment by LeeTibbert** [comment]
@ekrich @WojciechMazur   Did this get put to rest by recently merged PR #3758?

**Comment by WojciechMazur** [comment]
Thanks @LeeTibbert, good catch
